File: R/hashmap.R
#' @export
#' @method print Rcpp_rstruct_hashmap
print.Rcpp_rstruct_hashmap <- function(obj) {
  obj$print()
}

#' @export
#' @method [[ Rcpp_rstruct_hashmap
`[[.Rcpp_rstruct_hashmap` <- function(obj, key) {
  obj$get(key)
}

#' @export
#' @method [ Rcpp_rstruct_hashmap
`[.Rcpp_rstruct_hashmap` <- function(obj, keys) {
  obj$retrieve(keys)
}

#' @export
#' @method [[<- Rcpp_rstruct_hashmap
`[[<-.Rcpp_rstruct_hashmap` <- function(obj, key, value) {
  obj$insert(key, value)
  invisible(obj)
}

#' @export
hashmap <- function(input = NULL) {
  if (is.null(input)) {
      return(new(rstruct_hashmap))
  } 

  if (!is.list(input) || any(is.na(names(input))) || any(names(input) == "")) {
    stop("Input needs to be a named list, or missing.")
  }
  
  new(rstruct_hashmap, input)
}

File: src/hashmap.cpp
#include "hashmap.h"

#include <Rcpp.h>

#include <string>
#include <unordered_map>

RCPP_MODULE(hashmap_module) {
  Rcpp::class_<HashMap>("rstruct_hashmap")
      .constructor()
      .constructor<Rcpp::List>()
      .method("contains", &HashMap::contains)
      .method("insert", &HashMap::insert)
      .method("get", &HashMap::get)
      .method("update", &HashMap::update)
      .method("lookup", &HashMap::lookup)
      .method("retrieve", &HashMap::retrieve)
      .method("print", &HashMap::print);
}
